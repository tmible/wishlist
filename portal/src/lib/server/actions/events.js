/** @module События действий */

/**
 * Событие добавления действия
 * @constant {string}
 */
export const AddAction = {};
