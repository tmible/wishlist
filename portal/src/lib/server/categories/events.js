/** @module События категорий */

/**
 * Событие получения категорий
 * @constant {string}
 */
export const GetCategories = 'get categories';

/**
 * Событие добавления категории
 * @constant {string}
 */
export const AddCategory = 'add category';

/**
 * Событие обновления категории
 * @constant {string}
 */
export const UpdateCategory = 'update category';

/**
 * Событие удаления категории
 * @constant {string}
 */
export const DeleteCategory = 'delete category';
