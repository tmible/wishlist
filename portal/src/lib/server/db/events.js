/** @module События адаптера БД */

/**
 * Событие выполнения транзакции
 * @constant {string}
 */
export const RunTransaction = 'run transaction';

/**
 * Событие выполнения SQL выражения с проверкой авторизации
 * @constant {string}
 */
export const RunStatementAuthorized = 'run statement authorized';
