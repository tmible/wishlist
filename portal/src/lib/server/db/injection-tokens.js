/** @module Токены внедрения адаптера БД */

/**
 * Токен внедрения БД
 * @constant {string}
 */
export const Database = 'wishlist database';

/**
 * SQL выражение для посчёта изменений
 * @constant {string}
 */
export const ChangesStatement = 'changes statement';
