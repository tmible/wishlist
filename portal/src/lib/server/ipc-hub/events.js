/** @module События IPC хаба */

/**
 * Событие получения новых данных из сокета
 * @constant {{}}
 */
export const SocketData = {};

/**
 * Событие ошибки сокета
 * @constant {{}}
 */
export const SocketError = {};

/**
 * Событие проверки статуса сокета
 * @constant {{}}
 */
export const CheckSocketConnection = {};

/**
 * Событие отправки данных сокету
 * @constant {{}}
 */
export const WriteToSocket = {};
