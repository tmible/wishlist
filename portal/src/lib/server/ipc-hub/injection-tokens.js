/** @module Токены внедрения адаптера IPC хаба */

/**
 * Токен внедрения IPC хаба
 * @constant {string}
 */
export const IPCHub = 'IPC hub';
