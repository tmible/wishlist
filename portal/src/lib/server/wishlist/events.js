/** @module События списка желаний */

/**
 * Событие получения списка
 * @constant {string}
 */
export const GetWishlist = 'get wishlist';

/**
 * Событие добавления элемента к списку
 * @constant {string}
 */
export const AddItem = 'add item';

/**
 * Событие сохранения элементов разметки описания
 * @constant {string}
 */
export const InsertDescriptionEntities = 'insert description entities';

/**
 * Событие получения элемента
 * @constant {string}
 */
export const GetItem = 'get item';

/**
 * Событие получения функции, подсчитывающей количество элементов списка желаний,
 * принадлежащих пользователю, среди указанных
 * @constant {string}
 */
export const GetWishlistItemsCounter = 'get wishlist items counter';

/**
 * Событие переупорядочивания списка
 * @constant {string}
 */
export const ReorderWishlist = 'reorder wishlist';

/**
 * Событие удаления нескольких элементов из списка
 * @constant {string}
 */
export const DeleteItems = 'delete items';

/**
 * Событие обновления элемента списка
 * @constant {string}
 */
export const UpdateItem = 'update item';

/**
 * Событие удаления элементов разметки описания
 * @constant {string}
 */
export const DeleteDescriptionEntities = 'delete description entities';

/**
 * Событие удаления одного элемента из списка
 * @constant {string}
 */
export const DeleteItem = 'delete item';
