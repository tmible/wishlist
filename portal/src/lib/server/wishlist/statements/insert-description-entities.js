import DescriptionEntityBaseKeys from '@tmible/wishlist-common/constants/description-entity-base-keys';
import { inject } from '@tmible/wishlist-common/dependency-injector';
import { subscribe } from '@tmible/wishlist-common/event-bus';
import { Database } from '$lib/server/db/injection-tokens.js';
import { InsertDescriptionEntities } from '../events.js';

/** @typedef {import('$lib/components/telegram-entities/parser.svelte').Entity} Entity */

/**
 * Создание и выполнение SQL выражения для записи элементов разметки описания в БД
 * @function insertDescriptionEntities
 * @param {number | string} itemId Идентификатор элемента списка, к которому относится описание
 * @param {Entity[]} descriptionEntities Элементы разметки описания
 * @returns {void}
 */
const insertDescriptionEntities = (itemId, descriptionEntities) => {
  if (descriptionEntities.length <= 0) {
    return;
  }

  inject(Database).prepare(
    `INSERT INTO description_entities VALUES ${
      descriptionEntities.map((entity) => `($itemId, ?, ?, ?, ${
        Object.entries(entity).some(([ key ]) => !DescriptionEntityBaseKeys.has(key)) ?
          '?' :
          'NULL'
      })`).join(',')
    }`,
  ).run(
    ...descriptionEntities.flatMap((entity) => {
      const additionalProperties = Object.entries(entity).filter(
        ([ key ]) => !DescriptionEntityBaseKeys.has(key),
      );
      return [
        entity.type,
        entity.offset,
        entity.length,
        JSON.stringify(Object.fromEntries(additionalProperties)),
      ].slice(0, additionalProperties.length > 0 ? 4 : 3);
    }),
    { itemId },
  );
};

/**
 * Подписка [выполнения SQL выражения]{@link insertDescriptionEntities} на соответствующее событие
 * @function initInsertDescriptionEntitiesStatement
 * @returns {void}
 */
export const initInsertDescriptionEntitiesStatement = () => {
  subscribe(InsertDescriptionEntities, insertDescriptionEntities);
};
