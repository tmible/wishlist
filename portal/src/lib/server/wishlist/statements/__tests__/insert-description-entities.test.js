import DescriptionEntityBaseKeys from '@tmible/wishlist-common/constants/description-entity-base-keys';
import { inject } from '@tmible/wishlist-common/dependency-injector';
import { subscribe } from '@tmible/wishlist-common/event-bus';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { Database } from '$lib/server/db/injection-tokens.js';
import { InsertDescriptionEntities } from '../../events.js';
import { initInsertDescriptionEntitiesStatement } from '../insert-description-entities.js';

vi.mock('@tmible/wishlist-common/dependency-injector');
vi.mock('@tmible/wishlist-common/event-bus');
vi.mock('$lib/server/db/injection-tokens.js', () => ({ Database: 'database' }));
vi.mock('../../events.js', () => ({ InsertDescriptionEntities: 'insert description entities' }));

describe('wishlist / statements / insert description entities', () => {
  let db;
  let statement;

  beforeEach(() => {
    statement = { run: vi.fn() };
    db = { prepare: vi.fn(() => statement) };
    vi.mocked(inject).mockReturnValue(db);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should subscribe to event', () => {
    initInsertDescriptionEntitiesStatement();
    expect(
      vi.mocked(subscribe),
    ).toHaveBeenCalledWith(
      vi.mocked(InsertDescriptionEntities),
      expect.any(Function),
    );
  });

  describe('subscriber if there are entities', () => {
    const baseDescriptionEntity = Object.fromEntries(
      Array.from(DescriptionEntityBaseKeys, (key) => [ key, key ]),
    );

    let eventHandler;

    beforeEach(() => {
      vi.mocked(subscribe).mockImplementationOnce((event, handler) => eventHandler = handler);
      initInsertDescriptionEntitiesStatement();
    });

    it('should inject database', () => {
      eventHandler('userid', [ 1, 2, 3 ]);
      expect(vi.mocked(inject)).toHaveBeenCalledWith(vi.mocked(Database));
    });

    it('should prepare statement', () => {
      eventHandler(
        'id',
        [
          { ...baseDescriptionEntity, any: 'other' },
          baseDescriptionEntity,
        ],
      );
      expect(db.prepare.mock.calls[0]).toMatchSnapshot();
    });

    it('should run statement on event emit', () => {
      eventHandler(
        'id',
        [
          { ...baseDescriptionEntity, any: 'other' },
          baseDescriptionEntity,
        ],
      );
      expect(statement.run.mock.calls[0]).toMatchSnapshot();
    });
  });
});
