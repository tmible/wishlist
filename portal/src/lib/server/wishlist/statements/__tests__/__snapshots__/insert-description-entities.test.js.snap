// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wishlist / statements / insert description entities > subscriber if there are entities > should prepare statement 1`] = `
[
  "INSERT INTO description_entities VALUES ($itemId, ?, ?, ?, ?),($itemId, ?, ?, ?, NULL)",
]
`;

exports[`wishlist / statements / insert description entities > subscriber if there are entities > should run statement on event emit 1`] = `
[
  "type",
  "offset",
  "length",
  "{"any":"other"}",
  "type",
  "offset",
  "length",
  {
    "itemId": "id",
  },
]
`;
